src/main/java/io/fusionauth/samlv2/service/DefaultSAMLv2Service.java

```diff
@@ -874,6 +1023,148 @@ private void checkFor_CVE_2022_21449(SAMLRequest request, byte[] signature) thro
     }
   }
 
+  /**
+   * Creates the algorithm parameter spec according to the selected encryption algorithm
+   *
+   * @param encryptionAlgorithm The encryption algorithm
+   * @param iv                  The initialization vector
+   * @return The algorithm parameter spec for initializing the {@link Cipher}
+   */
+  private AlgorithmParameterSpec createAlgorithmParameterSpec(EncryptionAlgorithm encryptionAlgorithm, byte[] iv) {
+    if (List.of(EncryptionAlgorithm.AES128GCM, EncryptionAlgorithm.AES192GCM, EncryptionAlgorithm.AES256GCM).contains(encryptionAlgorithm)) {
+      return new GCMParameterSpec(128, iv);
+    } else {
+      return new IvParameterSpec(iv);
+    }
+  }
+
+  /**
+   * Encrypt the SAML Assertion in the XML document and return a new XML document with the Assertion replaced by
+   * EncryptedAssertion
+   *
+   * @param document              The XML document containing the SAML Response with an unencrypted Assertion
+   * @param encryptionAlgorithm   The algorithm used to encrypt the SAML Assertion
+   * @param keyLocation           The location to place the EncryptedKey in EncryptedAssertion
+   * @param transportAlgorithm    The algorithm used to encrypt the symmetric key for transport
+   * @param encryptionCertificate The certificate containing the public key for encrypting the symmetric key
+   * @param digest                The message digest algorithm to use with RSA-OAEP encryption (if necessary)
+   * @param mgf                   The mask generation function to use with RSA-OAEP encryption (if necessary)
+   * @return A new XML document containing the SAML response with an EncryptedAssertion
+   * @throws SAMLException if there is an issue encrypting the SAML Assertion or generating a new document
+   */
+  private Document encryptAssertion(Document document, EncryptionAlgorithm encryptionAlgorithm, KeyLocation keyLocation,
```

Comment:
This method contains most of the logic for encrypting the assertion. The other new methods just handle smaller parts of the task.